library/nameops.ml

```diff
@@ -124,34 +124,86 @@ let atompart_of_id id = fst (repr_ident id)
 
 (* Names *)
 
-let out_name = function
-  | Name id -> id
-  | Anonymous -> failwith "Nameops.out_name"
+module type ExtName =
+sig
 
-let name_fold f na a =
+  include module type of struct include Names.Name end
+
+  val fold_left : ('a -> Id.t -> 'a) -> 'a -> t -> 'a
+  val fold_right : (Id.t -> 'a -> 'a) -> t -> 'a -> 'a
+  val iter : (Id.t -> unit) -> t -> unit
+  val map : (Id.t -> Id.t) -> t -> t
+  val fold_map : ('a -> Id.t -> 'a * Id.t) -> 'a -> t -> 'a * t
+  val out : t -> Id.t
+  val max : t -> t -> t
+  val cons : t -> Id.t list -> Id.t list
+  val to_option : Name.t -> Id.t option
+
+end
+
+module Name =
+struct
+
+  include Names.Name
+
+  exception IsAnonymous
+
+  let fold_left f a na =
   match na with
-  | Name id -> f id a
+  | Name id -> f a id
   | Anonymous -> a
 
-let name_iter f na = name_fold (fun x () -> f x) na ()
-
-let name_cons na l =
+  let fold_right f na a =
   match na with
-  | Anonymous -> l
-  | Name id -> id::l
+  | Name id -> f id a
+  | Anonymous -> a
+
+  let iter f na = fold_right (fun x () -> f x) na ()
 
-let name_app f = function
+  let map f = function
   | Name id -> Name (f id)
   | Anonymous -> Anonymous
 
-let name_fold_map f e = function
-  | Name id -> let (e,id) = f e id in (e,Name id)
-  | Anonymous -> e,Anonymous
+  let fold_map f a = function
+  | Name id -> let (a, id) = f a id in (a, Name id)
+  | Anonymous -> a, Anonymous
+
+  let out = function
+    | Name id -> id
+    | Anonymous -> raise IsAnonymous
+
+  let max na1 na2 =
+    match na1 with
+    | Name _ -> na1
+    | Anonymous -> na2
```

Comment:
What does `max` mean here? I read it as maximum. But since this function does not induce any kind of lattice or order (as far as I can tell), my interpretation of the name is presumably incorrect. I understand this is inherited from the old name, but if the old name was indeed counterintuitive, we might want to change it in the new interface.